plugins/ti-dmc/fu-ti-dmc-device.c

```diff
@@ -0,0 +1,737 @@
+{
+	FuChunk *chk_old = NULL;
+
+	/* progress */
+	fu_progress_set_id(progress, G_STRLOC);
+	fu_progress_set_steps(progress, chunks->len);
+	for (guint i = 0; i < chunks->len; i++) {
+		FuChunk *chk = g_ptr_array_index(chunks, i);
+		g_autoptr(GByteArray) buf = g_byte_array_new();
+
+		/* align */
+		g_byte_array_append(buf, fu_chunk_get_data(chk), fu_chunk_get_data_sz(chk));
+		fu_byte_array_align_up(buf, FU_FIRMWARE_ALIGNMENT_64, 0xFF);
+
+		/* ignore duplicate chunk */
+		if (chk_old != NULL && fu_chunk_get_data_sz(chk) > 4) {
+			if (fu_memcmp_safe(fu_chunk_get_data(chk),
+					   fu_chunk_get_data_sz(chk),
+					   fu_chunk_get_data(chk_old),
+					   fu_chunk_get_data_sz(chk_old),
+					   NULL)) {
+				g_debug("skipping duplicate chunk %u", fu_chunk_get_idx(chk));
+				continue;
+			}
+		}
+
+		/* FIXME: this is retried twice on failure -- should we do the same thing? */
+		if (!fu_ti_dmc_device_sfwd(self, buf, error))
+			return FALSE;
+
+		/* save for comparison */
+		if (fu_chunk_get_data_sz(chk) > 4)
+			chk_old = chk;
+
+		/* update progress */
+		g_usleep(100 * 1000);
+		fu_progress_step_done(progress);
+	}
+
+	/* success */
+	return TRUE;
+}
+
+static gboolean
+fu_ti_dmc_device_write_sfws_chunks(FuTiDmcDevice *self,
+				   GPtrArray *chunks,
+				   FuProgress *progress,
+				   GError **error)
+{
+	/* progress */
+	fu_progress_set_id(progress, G_STRLOC);
+	fu_progress_set_steps(progress, chunks->len);
+	for (guint i = 0; i < chunks->len; i++) {
+		FuChunk *chk = g_ptr_array_index(chunks, i);
+		g_autoptr(GByteArray) buf = g_byte_array_new();
+
+		/* align and pad low before sending */
+		g_byte_array_append(buf, fu_chunk_get_data(chk), fu_chunk_get_data_sz(chk));
+		fu_byte_array_align_up(buf, FU_FIRMWARE_ALIGNMENT_64, 0x0);
+		if (!fu_ti_dmc_device_sfws(self, buf, error))
+			return FALSE;
+
+		/* update progress */
+		fu_progress_step_done(progress);
+	}
+
+	/* success */
+	return TRUE;
+}
+
+static gboolean
+fu_ti_dmc_device_write_firmware(FuDevice *device,
+				FuFirmware *firmware,
+				FuProgress *progress,
+				FwupdInstallFlags flags,
+				GError **error)
+{
+	FuTiDmcDevice *self = FU_TI_DMC_DEVICE(device);
+	g_autoptr(GBytes) fw = NULL;
+	g_autoptr(GBytes) fw_sig = NULL;
+	g_autoptr(GBytes) fw_pubkey = NULL;
+	g_autoptr(GBytes) fw_payload = NULL;
+	g_autoptr(GPtrArray) chunks_pubkey = NULL;
+	g_autoptr(GPtrArray) chunks_sig = NULL;
+	g_autoptr(GPtrArray) chunks_payload = NULL;
+
+	/* progress */
+	fu_progress_set_id(progress, G_STRLOC);
+	fu_progress_add_flag(progress, FU_PROGRESS_FLAG_GUESSED);
+	fu_progress_add_step(progress, FWUPD_STATUS_DEVICE_BUSY, 1, NULL);
+	fu_progress_add_step(progress, FWUPD_STATUS_DEVICE_WRITE, 19, NULL);
+	fu_progress_add_step(progress, FWUPD_STATUS_DEVICE_WRITE, 44, NULL);
+	fu_progress_add_step(progress, FWUPD_STATUS_DEVICE_WRITE, 44, NULL);
+
+	/* get payload image */
+	fw_payload = fu_firmware_get_image_by_id_bytes(firmware, FU_FIRMWARE_ID_PAYLOAD, error);
+	if (fw_payload == NULL)
```

Comment:
Hm... here I have `NULL` while trying to use `install-blob` command with FWs I have.

Reply:
oopa, can you try with the commit I just pushed -- we're not setting the firmware type on the device.

Reply:
> and segfault after

Can you try the new commit -- I'm forgetting to create the g_byte_array... :/

Reply:
Sure, now have `response 0x7 but requested 0x40`.
on it.

Reply:
So in the original code, read_reg says it always returns 64 bytes. Maybe the test should be `if (buf->data[0] < length)`? I think fwupd is being more strict that the original code.

Reply:
next step `SFWs failed`

Reply:
Is it a bit weird that "SFWs with signature" and "SFWs with pubkey" both call fu_ti_dmc_device_write_sfws_chunks?

Reply:
Maybe not so weird. I tried myself and pushed what we have so far, and a few lines to add proper error messages. Now we have:

    failed to write SFWs with signature: failed to write chunk 0: SFWs failed, got sfwd-not-run-or-no-key-exists [0x06]

you see the same?

Reply:
Also, debugging this a bit more I see:

	36 5a 02 20 ff 00 00 00 00 80 30 5a 02 20 79 42 5a 02 20 ff 00 00 00 00 01 00 00 ef 00 00 00 00 
	00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 
	15:17:55.489 FuPluginTiDmc        write@0x1e:     53 46 57 64 
	15:17:55.490 FuPluginTiDmc        read@0x1e:      04 53 46 57 64 
	15:17:55.591 FuPluginTiDmc        read@0x1e:      04 00 00 00 00 
	15:17:55.592 FuPluginTiDmc        read@0x1f:      40 80 
	15:17:55.592 FuPluginTiDmc        more-data-expected: 1
	15:17:55.692 FuPluginTiDmc        write@0x1f:
	00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 
	00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 
	15:17:55.699 FuPluginTiDmc        write@0x1e:     53 46 57 64 
	15:17:55.700 FuPluginTiDmc        read@0x1e:      04 53 46 57 64 
	15:17:55.801 FuPluginTiDmc        read@0x1e:      04 00 00 00 00 
	15:17:55.802 FuPluginTiDmc        read@0x1f:      40 80 
	15:17:55.802 FuPluginTiDmc        more-data-expected: 1
	15:17:55.902 FuPluginTiDmc        write@0x1f:
	00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 
	00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 ff ff ff ff 
	15:17:55.909 FuPluginTiDmc        write@0x1e:     53 46 57 64 
	15:17:55.910 FuPluginTiDmc        read@0x1e:      04 53 46 57 64 
	15:17:56.011 FuPluginTiDmc        read@0x1e:      04 00 00 00 00 
	15:17:56.012 FuPluginTiDmc        read@0x1f:      40 80 
	15:17:56.012 FuPluginTiDmc        more-data-expected: 1

i.e. we're writing all the SFWd blocks, and even padding out the last block with `0xFF` like we expect. But we still get the `more-data-expected: 1` -- I did kinda expect that to be false for the last block.

Reply:
not exactly.
In my case the data from common_update is written and after that I've got a failure:

```
14:34:40.327 FuPluginTiDmc        SFWd: 0x80, 0x97
14:34:40.428 FuPluginTiDmc        Write@0x1f:
00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00
00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 ff ff ff ff
14:34:40.433 FuPluginTiDmc        Write@0x1e:     53 46 57 64
14:34:40.434 FuPluginTiDmc        Read@0x1e:      04 53 46 57 64
14:34:40.534 FuPluginTiDmc        Read@0x1e:      04 00 00 00 00
14:34:40.536 FuPluginTiDmc        Read@0x1f:
       | 00 01 02 03 04 05 06 07 08 09 0a 0b 0c 0d 0e 0f 10 11 12 13 14 15 16 17 18 19 1a 1b 1c 1d 1e 1f
-------+------------------------------------------------------------------------------------------------
0x0000 | 40 80 98 02 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00
0x0020 | 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 ff ff ff
0x0040 | ff
14:34:40.536 FuPluginTiDmc        SFWd: 0x80, 0x98
14:34:40.636 FuProgress           child is at 664/750 step_max and parent done [../plugins/ti-dmc/fu-ti-dmc-device.c:642]
4) ../src/fu-tool.c:892 (2/3)
3) ../src/fu-engine.c:3454 (1/3)
2) ../plugins/ti-dmc/fu-ti-dmc-device.c:708 (1/4)
1) ../plugins/ti-dmc/fu-ti-dmc-device.c:642 (1/4)
0) ../plugins/ti-dmc/fu-ti-dmc-device.c:559 (664/750)
  
14:34:40.637 FuPluginTiDmc        Write@0x1f:
27 a8 fa 49 46 1f 07 ae 32 75 41 3a 0f d1 88 1d 34 54 05 58 01 7d f3 b2 44 ec e0 43 ad 7f 80 f5
57 3e 46 c7 3a 14 95 87 98 2d 32 0b 78 67 b8 50 5d 01 b8 32 6c cd 99 ee b5 39 c4 69 af b3 a6 86
14:34:40.641 FuPluginTiDmc        Write@0x1e:     53 46 57 73
14:34:40.642 FuPluginTiDmc        Read@0x1e:      04 00 00 00 00
14:34:40.644 FuPluginTiDmc        Read@0x1f:
       | 00 01 02 03 04 05 06 07 08 09 0a 0b 0c 0d 0e 0f 10 11 12 13 14 15 16 17 18 19 1a 1b 1c 1d 1e 1f
-------+------------------------------------------------------------------------------------------------
0x0000 | 40 06 00 00 00 46 1f 07 ae 32 75 41 3a 0f d1 88 1d 34 54 05 58 01 7d f3 b2 44 ec e0 43 ad 7f 80
0x0020 | f5 57 3e 46 c7 3a 14 95 87 98 2d 32 0b 78 67 b8 50 5d 01 b8 32 6c cd 99 ee b5 39 c4 69 af b3 a6
0x0040 | 86
14:34:40.644 FuPluginTiDmc        SFWs: 0x6, 0x0
14:34:40.646 FuPlugin             attach(ti_dmc)
14:34:40.646 FuDeviceList         no replug or re-enumerate required
14:34:40.646 FuEngine             cleanup -> FuTiDmcDevice:

14:34:40.646 FuDeviceList         no replug or re-enumerate required
SFWs failed, got 0x06
```
Now checking the original workflow in python to compare from scratch.

Reply:
On a [risky?] hunch I disabled the ignoring of duplicate chunks, and how I get `more-data-expected: 0` and:

15:31:43.430 FuPluginTiDmc        read@0x1e:      04 00 00 00 00
	15:31:43.430 FuPluginTiDmc        read@0x1f:      40 00 0c 36 00 00 00 00 00 00 00
	15:31:43.430 FuPluginTiDmc        more-data-expected: 0
	15:31:43.430 FuPluginTiDmc        signature-data-block: 12
	15:31:43.430 FuPluginTiDmc        prod-key-present: 2
	15:31:43.430 FuPluginTiDmc        engr-key-present: 4
	15:31:43.430 FuPluginTiDmc        new-flash-region: 16

...which I *think* looks like a successful flash!

The device seems to reboot after calling fu_ti_dmc_device_reset_hard() -- which is maybe expected!

Reply:
argh, I didn't pull last changes -- that's why....